# Data Wrangling Template

## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess
These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)

Try them out below and keep their results in mind. Some will come in handy later in the lesson.

Check out the [pandas API reference](https://pandas.pydata.org/pandas-docs/stable/api.html) for detailed usage information.

In [3]:
patients.head(n=10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   
5      Daytona Beach     Florida   32114.0  United States   
6            Burbank          NV   84728.0  United States   
7        Morgan Hill          CA   95037.0  United States   
8        Kansas City          MO   64105.0  United States   
9           New York    New York   10011.0  United States   

                                         contact   birthdate  weight  height  \
0           951-719-9170ZoeWellish@superrito.com   7/10/1976   121.7      66   
1          PamelaSHill@cuvox.de+1 (217) 569-3204    4/3/1967   118.8      66   
2               402-363-6804JaeMDebord@gustr.com   2/19/1980   177.8      71   
3     PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
4                334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
5       386-334-5237RafaelCardosoCosta@gustr.com   8/31/1931   183.9      70   
6              775-533-5933MaryBAdams@einrot.com  11/19/1969   146.3      65   
7           XiuxiuChang@einrot.com1 408 778 3236   8/13/1958   158.0      60   
8        816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   
9  SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7      64   

    bmi  
0  19.6  
1  19.2  
2  24.8  
3  31.7  
4  26.1  
5  26.4  
6  24.3  
7  30.9  
8  26.5  
9  33.4

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [4]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [5]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [6]:
patients.state.value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
OK            13
MI            13
Florida       13
LA            13
NJ            12
VA            11
WI            10
IL            10
MS            10
AL             9
IN             9
FL             9
MN             9
TN             9
NC             8
WA             8
KY             8
MO             7
ID             6
NV             6
KS             6
CT             5
IA             5
SC             5
CO             4
AR             4
RI             4
ME             4
ND             4
AZ             4
Nebraska       4
WV             3
DE             3
OR             3
MD             3
SD             3
DC             2
NE             2
VT             2
MT             2
AK             1
NM             1
NH             1
WY             1
Name: state, dtype: int64

In [7]:
patients[patients.city=="New York"]

patient_id assigned_sex    given_name     surname               address  \
9            10       female        Sophie     Cabrera     3303 Anmoore Road   
35           36       female        Kamila    Pecinová  3558 Longview Avenue   
84           85       female         Nương          Vũ   465 Southern Street   
129         130       female       Rebecca    Jephcott      989 Wayback Lane   
142         143         male        Finley    Chandler  2754 Westwood Avenue   
152         153         male   Christopher    Woodward  3450 Southern Street   
188         189         male         Søren    Sørensen      2397 Bell Street   
213         214       female    Onyemaechi   Onwughara     685 Duncan Avenue   
215         216         male          John         Doe       123 Main Street   
229         230         male          John         Doe       123 Main Street   
237         238         male          John         Doe       123 Main Street   
244         245         male          John         Doe       123 Main Street   
247         248         male        Tuukka  Leppäluoto    1886 Bicetown Road   
251         252         male          John         Doe       123 Main Street   
263         264       female         Julia    Carvalho     3662 Shinn Street   
277         278         male          John         Doe       123 Main Street   
301         302       female  Onyekachukwu      Obinna    2970 Forest Avenue   
461         462         male        Cannan     Cabrera   2102 Geraldine Lane   

         city     state  zip_code        country  \
9    New York  New York   10011.0  United States   
35   New York  New York   10004.0  United States   
84   New York        NY   10001.0  United States   
129  New York        NY   10004.0  United States   
142  New York  New York   10001.0  United States   
152  New York        NY   10004.0  United States   
188  New York        NY   10011.0  United States   
213  New York        NY   10013.0  United States   
215  New York        NY   12345.0  United States   
229  New York        NY   12345.0  United States   
237  New York        NY   12345.0  United States   
244  New York        NY   12345.0  United States   
247  New York        NY   10011.0  United States   
251  New York        NY   12345.0  United States   
263  New York        NY   10036.0  United States   
277  New York        NY   12345.0  United States   
301  New York        NY   10004.0  United States   
461  New York        NY   10014.0  United States   

                                               contact   birthdate  weight  \
9        SophieCabreraIbarra@teleworm.us1 718 795 9124   12/3/1930   194.7   
35               718-501-0503KamilaPecinova@dayrep.com  12/23/1985   198.9   
84                  VuCamNuong@fleckens.hu516-720-5094    2/1/1981   138.2   
129            631-370-7406RebeccaJephcott@armyspy.com    8/1/1966   203.3   
142              516-740-5280FinleyChandler@dayrep.com  10/25/1936   150.9   
152  ChristopherWoodward@jourrapide.com+1 (516) 630...    9/4/1984   212.2   
188            SrenSrensen@superrito.com1 212 201 3108  12/31/1942   157.1   
213         917-622-9142OnyemaechiOnwughara@einrot.com    3/8/1989   131.1   
215                        johndoe@email.com1234567890    1/1/1975   180.0   
229                        johndoe@email.com1234567890    1/1/1975   180.0   
237                        johndoe@email.com1234567890    1/1/1975   180.0   
244                        johndoe@email.com1234567890    1/1/1975   180.0   
247           917-408-8855TuukkaLeppaluoto@teleworm.us    3/7/1978   211.0   
251                        johndoe@email.com1234567890    1/1/1975   180.0   
263  JuliaAzevedoCarvalho@superrito.com+1 (212) 782...   4/11/1931   171.8   
277                        johndoe@email.com1234567890    1/1/1975   180.0   
301         OnyekachukwuObinna@teleworm.us646-982-6609   1/24/1997   154.7   
461     646-289-4177CannanCabreraOrdonez@superrito.com  10/12/1980   209.7   

     height 

In [8]:
##bmi formula, check for deviations.
patients[abs((703*patients.weight/(patients.height**2))-patients.bmi)>.05]

patient_id assigned_sex given_name   surname               address  \
4             5         male        Tim   Neudorf  1428 Turkey Pen Lane   
210         211       female    Camilla  Zaitseva   4689 Briarhill Lane   

        city state  zip_code        country  \
4     Dothan    AL   36303.0  United States   
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
4              334-515-7487TimNeudorf@cuvox.de   2/18/1928   192.3      27   
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
4    26.1  
210  19.1

In [9]:
patients.surname.value_counts()

Doe                  6
Jakobsen             3
Taylor               3
Souza                2
Collins              2
Correia              2
Berg                 2
Kadyrov              2
Johnson              2
Dratchev             2
Cindrić              2
Liễu                 2
Tạ                   2
Silva                2
Parker               2
Bùi                  2
Grímsdóttir          2
Aranda               2
Nilsen               2
Ogochukwu            2
Tucker               2
Gersten              2
Lương                2
Lund                 2
Batukayev            2
Schiavone            2
Woźniak              2
Cabrera              2
Lâm                  2
Kowalczyk            2
                    ..
Chidi                1
Fomin                1
Pecha                1
Hsueh                1
Barsukov             1
Yao                  1
Scheltens            1
Yudina               1
Cain                 1
Brodahl              1
Kristiansen          1
Cumpston             1
de Keizer  

In [10]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2778 North Avenue             2
2476 Fulton Street            2
2403 Ripple Street            1
4271 Cherry Ridge Drive       1
1990 Spring Avenue            1
1463 Beechwood Avenue         1
3402 Kildeer Drive            1
1846 Joseph Street            1
2970 Forest Avenue            1
570 Alpha Avenue              1
98 Freshour Circle            1
4155 Raccoon Run              1
4649 Joanne Lane              1
541 Green Hill Road           1
1540 Overlook Drive           1
4273 Deans Lane               1
4992 Moonlight Drive          1
4220 Simpson Square           1
2931 Romano Street            1
1733 Blackwell Street         1
1403 Clousson Road            1
2691 Kessla Way               1
4458 Stark Hollow Road        1
2458 Broadway Avenue          1
1051 Bassell Avenue           1
2386 Linda Street             1
1975 Holden Street            1
1262 Deans Lane               1
                             ..
4160 Pra

In [11]:
patients[patients.address.duplicated(keep=False)]

patient_id assigned_sex given_name      surname             address  \
24           25         male      Jakob     Jakobsen   648 Old Dear Lane   
29           30         male       Jake     Jakobsen   648 Old Dear Lane   
97           98         male    Patrick      Gersten   2778 North Avenue   
131         132       female     Sandra       Taylor  2476 Fulton Street   
209         210       female     Lalita  Eldarkhanov                 NaN   
215         216         male       John          Doe     123 Main Street   
219         220         male         Mỹ        Quynh                 NaN   
229         230         male       John          Doe     123 Main Street   
230         231       female  Elisabeth      Knudsen                 NaN   
234         235       female    Martina    Tománková                 NaN   
237         238         male       John          Doe     123 Main Street   
242         243         male       John      O'Brian                 NaN   
244         245         male       John          Doe     123 Main Street   
249         250         male   Benjamin       Mehler                 NaN   
251         252         male       John          Doe     123 Main Street   
257         258         male        Jin         Kung                 NaN   
264         265       female   Wafiyyah       Asfour                 NaN   
269         270       female     Flavia   Fiorentino                 NaN   
277         278         male       John          Doe     123 Main Street   
278         279       female   Generosa        Cabán                 NaN   
282         283       female      Sandy       Taylor  2476 Fulton Street   
286         287         male      Lewis         Webb                 NaN   
296         297       female        Chỉ          Lâm                 NaN   
502         503         male        Pat      Gersten   2778 North Avenue   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
97          Burr        NE   68324.0  United States   
131     Rainelle        WV   25962.0  United States   
209          NaN       NaN       NaN            NaN   
215     New York        NY   12345.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
97         PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   
131        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
209                                         NaN   8/14/1950   143.4      62   
215                 johndoe@email.com1234567890    1/1/1975   180.0      72   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230              

In [12]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [15]:
sum(treatments.auralin.isnull())

0

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [16]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [17]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [18]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
nausea                        2
cough                         2
throat irritation             2
Name: adverse_reaction, dtype: int64

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

## Data Quality Dimensions:
Data quality dimensions help guide your thought process while assessing and also cleaning. The four main data quality dimensions are:

- Completeness: do we have all of the records that we should? Do we have missing records or not? Are there specific rows, columns, or cells missing?
- Validity: we have the records, but they're not valid, i.e., they don't conform to a defined schema. A schema is a defined set of rules for data. These rules can be real-world constraints (e.g. negative height is impossible) and table-specific constraints (e.g. unique key constraints in tables).
- Accuracy: inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect. Example: a patient's weight that is 5 lbs too heavy because the scale was faulty.
- Consistency: inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing. Consistency, i.e., a standard format, in columns that represent the same data across tables and/or within tables is desired.

### Quality issues

#### `patients` table:
- zip code is a float not a string
- zip code has four digits sometimes, not 5
- Tim Neudorf height is 27in, instead of 72in
- Camilla Zaitseva weight is 48, based on bmi should be 108 (in kg instead of lbs)
- some states have full name, some abbreviations.
- contacts has both phone number and email in one column
- patient_id 9 first name of Dsvid instead of David
- missing demographic information (address thorugh country, contact)
- state should be category instead of object type
- assigned sex should be category instead of object type
- birthdate should be datetime not object
- multiple phone number formats
- there are duplicate default john doe records
- duplication of jakob/jake jakobsen, patrick/pat gersten, and sandra/sandy taylor. Need to remove correct one.

#### `treatments` table: 
- missing HbA1c changes
- only has 280 rows, out of 350 participants
- u next to start and end dose (auralin and novodra should be int not objects
- given_name and surname are lowercase
- HbA1c values should be the same as start-end
- auralin and novodra null values are misrepresented (due to string type)

#### `adverse_reactions` table:
- given_name and surname are lowercase


 

### Tidyness issues

#### `patients` table:
- two variables in the contacts column, split into phone and email

#### `treatments` table:
- auralin and novodra columns need to be three columns: start dose, end dose, and treatment type (which is in the column headers)

#### `adverse_reactions` table:
- can be combined into treatments table


## Clean

1. fix completeness
2. fix tidyness

In [19]:
patients_clean=patients.copy()
treatments_clean=treatments.copy()
adverse_reactions_clean=adverse_reactions.copy()

#### Define

- `treatments`: only has 280 rows, out of 350 participants

Join missing treatment records from treatments_cut.csv into `treatments_clean` table

#### Code

In [20]:
treatments_cut=pd.read_csv('treatments_cut.csv')
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
given_name      70 non-null object
surname         70 non-null object
auralin         70 non-null object
novodra         70 non-null object
hba1c_start     70 non-null float64
hba1c_end       70 non-null float64
hba1c_change    42 non-null float64
dtypes: float64(3), object(4)
memory usage: 3.9+ KB


In [21]:
treatments_clean=pd.concat([treatments_clean,treatments_cut],ignore_index=True)

#### Test

In [22]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


#### Define

- `treatments`: missing/inaccurate HbA1c_change values

Fill in HbA1c column by subtracting HbA1c_end from HbA1c_start, replace all values.

#### Code

In [23]:
treatments_clean.hba1c_change=treatments_clean.hba1c_start-treatments_clean.hba1c_end


#### Test

In [24]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

#### Define


 - Contact column in `patients` table contains two variables: phone number and email
 
 Use Regex to split this column into two.
 
help sources: regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). [regex tutorial](https://regexone.com/). [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). [email address regex pattern](http://emailregex.com/).

#### Code

In [25]:
patients_clean.contact.head()

0          951-719-9170ZoeWellish@superrito.com
1         PamelaSHill@cuvox.de+1 (217) 569-3204
2              402-363-6804JaeMDebord@gustr.com
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246
4               334-515-7487TimNeudorf@cuvox.de
Name: contact, dtype: object

In [26]:
patients_clean['phonenumber']=patients_clean.contact.str.extract('(\d?\s*\d{3}[-\.\s]\d{3}[-\.\s]\d{4}|\d?\s*\(\d{3}\)\s*\d{3}[-\.\s]\d{4}|\d{3}[-\.\s]\d{4}|\d{10})',expand=True)
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phonenumber  
0  19.6      951-719-9170  
1  19.2  1 (217) 569-3204  
2  24.8      402-363-6804  
3  31.7  1 (732) 636-8246  
4  26.1      334-515-7487

In [27]:
patients_clean['email']=patients_clean.contact.str.extract("([\D.+-]+@[\D]+\.[a-z]+)",expand=True)
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phonenumber                      email  
0  19.6      951-719-9170   ZoeWellish@superrito.com  
1  19.2  1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8      402-363-6804       JaeMDebord@gustr.com  
3  31.7  1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1      334-515-7487        TimNeudorf@cuvox.de

#### Test

In [28]:
patients_clean[patients_clean.phonenumber.isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi phonenumber  \
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2         NaN   
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1         NaN   
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4         NaN   
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2         NaN   
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4         NaN   
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6         NaN   
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2         NaN   
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1         NaN   
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1         NaN   
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4         NaN   
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6         NaN   
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1         NaN   

    email  
209   NaN  
219   NaN  
230   NaN  
234   NaN  
242   NaN  
249   NaN  
257   NaN  
264   NaN  
269   NaN  
278   NaN  
286   NaN  
296   NaN

In [29]:
patients_clean[['contact','phonenumber','email']].sample(10)

contact       phonenumber  \
261          JeremyMontagu@teleworm.us617-297-0387      617-297-0387   
97            PatrickGersten@rhyta.com402-848-4923      402-848-4923   
168   650-849-6900GalitCasarezBrito@jourrapide.com      650-849-6900   
378            FurutaOsman@rhyta.com1 901 316 3780    1 901 316 3780   
48               AnnikaVaara@rhyta.com312-719-7238      312-719-7238   
330      VascovandeWiel@rhyta.com+1 (843) 368-5129  1 (843) 368-5129   
9    SophieCabreraIbarra@teleworm.us1 718 795 9124    1 718 795 9124   
420            631-479-8171AdlanShishani@gustr.com      631-479-8171   
402         601-885-6550ManouckWubbels@armyspy.com      601-885-6550   
51       NoeArandaGranado@dayrep.com1 309 912 9553    1 309 912 9553   

                                email  
261         JeremyMontagu@teleworm.us  
97           PatrickGersten@rhyta.com  
168  GalitCasarezBrito@jourrapide.com  
378             FurutaOsman@rhyta.com  
48              AnnikaVaara@rhyta.com  
330          VascovandeWiel@rhyta.com  
9     SophieCabreraIbarra@teleworm.us  
420           AdlanShishani@gustr.com  
402        ManouckWubbels@armyspy.com  
51        NoeArandaGranado@dayrep.com

In [30]:
patients_clean = patients_clean.drop('contact', axis=1)

#### Define

- `treatments`: auralin and novodra columns need to be three columns: start dose, end dose, and treatment type (which is in the column headers)

Use melt and str.split() to split the columns.

help sources: pandas [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

#### Code

In [31]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [32]:
treatments_clean=pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1     elliot  richardson         7.56       7.09          0.47   auralin   
2   yukitaka    takenaka         7.68       7.25          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
4     alissa      montez         7.78       7.46          0.32   auralin   

        dose  
0  41u - 48u  
1          -  
2          -  
3  33u - 36u  
4          -

In [33]:
treatments_clean=treatments_clean[treatments_clean.dose!="-"]
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 698
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
treatment       350 non-null object
dose            350 non-null object
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


In [34]:
treatments_clean['dose_start'], treatments_clean['dose_end']=treatments_clean.dose.str.split('-',1).str
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose dose_start dose_end  
0  41u - 48u       41u       48u  
3  33u - 36u       33u       36u  
6  37u - 42u       37u       42u  
7  31u - 38u       31u       38u  
9  30u - 36u       30u       36u

In [35]:
treatments_clean = treatments_clean.drop('dose', axis=1)

#### Test

In [36]:
treatments_clean.sample(10)

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
642     annika        vaara         7.73       7.34          0.39   novodra   
286       sara        miles         7.53       7.12          0.41   auralin   
109     marius       hansen         7.97       7.53          0.44   auralin   
141     johana         hrdá         7.79       7.49          0.30   auralin   
209       kári   hervinsson         8.09       7.66          0.43   auralin   
685       mary        adams         7.65       7.26          0.39   novodra   
483      vasco  van de wiel         7.94       7.51          0.43   novodra   
160     harley       tucker         7.97       7.64          0.33   auralin   
163      malik      vaneker         7.55       7.12          0.43   auralin   
229     sandra       taylor         7.84       7.49          0.35   auralin   

    dose_start dose_end  
642       20u       21u  
286       30u       35u  
109       41u       47u  
141       54u       54u  
209       37u       43u  
685       32u       33u  
483       31u       29u  
160       47u       57u  
163       44u       54u  
229       51u       58u

#### Define

- `adverse_reactions` table: can be combined into treatments table

Join `adverse_reactions` and `treatments` tables using the merge function

In [37]:
adverse_reactions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


#### Code

In [38]:
treatments_clean=treatments_clean.merge(adverse_reactions_clean,how='left',on=['given_name','surname'])

#### Test

In [39]:
treatments_clean[treatments_clean.adverse_reaction.notnull()].sample(10)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
93         merci      leroux         8.98       8.64          0.34   auralin   
46        idalia       moore         8.34       7.90          0.44   auralin   
148       flavia  fiorentino         7.95       7.60          0.35   auralin   
178      jasmine       sykes         7.56       7.18          0.38   novodra   
175       elliot  richardson         7.56       7.09          0.47   novodra   
345  christopher    woodward         7.51       7.06          0.45   novodra   
119      manouck     wubbels         7.66       7.40          0.26   auralin   
86      gabriele     saenger         8.35       7.97          0.38   auralin   
227         abel     yonatan         7.88       7.50          0.38   novodra   
225    abdul-nur         isa         7.98       7.53          0.45   novodra   

    dose_start dose_end   adverse_reaction  
93        27u       33u       hypoglycemia  
46        25u       31u       hypoglycemia  
148       46u       57u              cough  
178       42u       44u       hypoglycemia  
175       40u       45u       hypoglycemia  
345       55u       51u             nausea  
119       55u       62u  throat irritation  
86        22u       30u       hypoglycemia  
227       38u       39u              cough  
225       54u       50u       hypoglycemia

In [40]:
treatments_clean.sample(10,random_state=10)

given_name    surname  hba1c_start  hba1c_end  hba1c_change treatment  \
43    stefanie     herman         7.72       7.30          0.42   auralin   
306     csilla   herczegh         7.71       7.27          0.44   novodra   
138  inunnguaq   heilmann         7.85       7.45          0.40   auralin   
275       lena       baer         7.70       7.40          0.30   novodra   
65        nora     nyborg         7.83       7.48          0.35   auralin   
6      roxanne  andreyeva         9.54       9.14          0.40   auralin   
262       kong        lei         7.58       7.15          0.43   novodra   
172     rovzan    kishiev         7.75       7.41          0.34   auralin   
342    bjørnar     nilsen         7.99       7.70          0.29   novodra   
218       else   andersen         7.98       7.60          0.38   novodra   

    dose_start dose_end adverse_reaction  
43        44u       55u              NaN  
306       43u       46u              NaN  
138       57u       67u              NaN  
275       41u       38u     hypoglycemia  
65        55u       59u              NaN  
6         29u       38u              NaN  
262       32u       30u              NaN  
172       32u       37u              NaN  
342       36u       33u              NaN  
218       36u       38u              NaN

#### Define

- `treatments` table: given_name and surname duplicated (should only be on `patients` table).
- `treatments` table: needs to have patient_ids.
 
merge patient_ids into the treatments table using merge, and the drop the given_name and surname columns.

#### Code

In [41]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names['given_name'] = id_names.given_name.str.lower()
id_names['surname'] = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/Users/devinmccormack/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/Users/devinmccormack/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


#### Test

In [42]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin       41u       48u   
1         7.97       7.62          0.35   auralin       33u       36u   
2         7.65       7.27          0.38   auralin       37u       42u   
3         7.89       7.55          0.34   auralin       31u       38u   
4         7.76       7.37          0.39   auralin       30u       36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [43]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

#### Define

- `patients` table: zip code is a float not a string, and sometimes has 4 digits instead of 5

coerce into string, remove .0, and prepend 0s when necessary with str.pad

#### Code

In [44]:
patients_clean.zip_code=patients_clean.zip_code.astype(str).str[:-2].str.pad(5,fillchar='0')

#### Test

In [45]:
patients_clean[['city','state','zip_code']].head()

city       state zip_code
0  Rancho California  California    92390
1          Armstrong    Illinois    61812
2               York    Nebraska    68467
3         Woodbridge          NJ    07095
4             Dothan          AL    36303

#### Define

- Tim Neudorf height is 27 in instead of 72 in

replace value with 72.

#### Code

In [46]:
patients_clean.height.loc[patients_clean.surname=='Neudorf']=72


/Users/devinmccormack/anaconda/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


#### Test

In [47]:
patients_clean.height.describe()

count    503.000000
mean      66.723658
std        4.047179
min       59.000000
25%       63.000000
50%       67.000000
75%       70.000000
max       79.000000
Name: height, dtype: float64

#### Define

- Full state names sometimes, abbreviations other times

replace all full names with abbreviations

#### Code

In [48]:
patients_clean.state.value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
OK            13
MI            13
Florida       13
LA            13
NJ            12
VA            11
WI            10
IL            10
MS            10
AL             9
IN             9
FL             9
MN             9
TN             9
NC             8
WA             8
KY             8
MO             7
ID             6
NV             6
KS             6
CT             5
IA             5
SC             5
CO             4
AR             4
RI             4
ME             4
ND             4
AZ             4
Nebraska       4
WV             3
DE             3
OR             3
MD             3
SD             3
DC             2
NE             2
VT             2
MT             2
AK             1
NM             1
NH             1
WY             1
Name: state, dtype: int64

In [49]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

#### Test

In [50]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States  7/10/1976   121.7      66   
1          Armstrong    IL    61812  United States   4/3/1967   118.8      66   
2               York    NE    68467  United States  2/19/1980   177.8      71   
3         Woodbridge    NJ    07095  United States  7/26/1951   220.9      70   
4             Dothan    AL    36303  United States  2/18/1928   192.3      72   

    bmi       phonenumber                      email  
0  19.6      951-719-9170   ZoeWellish@superrito.com  
1  19.2  1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8      402-363-6804       JaeMDebord@gustr.com  
3  31.7  1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1      334-515-7487        TimNeudorf@cuvox.de

#### Define

Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

#### Code

In [51]:
patients_clean.given_name=patients_clean.given_name.replace('Dsvid','David')


#### Test

In [52]:
patients_clean[patients_clean.surname=='Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

    phonenumber                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Define

- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

Convert set and state to categorical, birthdate to datetime. zip_code was taken care of above.

strip u from the start_dose and end_dose, and convert to numeric.

#### Code

In [53]:
#category
patients_clean.assigned_sex=patients_clean.assigned_sex.astype('category')
patients_clean.state=patients_clean.state.astype('category')

# datetime
patients_clean.birthdate=pd.to_datetime(patients_clean.birthdate)


In [54]:
# strip u

treatments_clean.dose_start=treatments_clean.dose_start.str.replace('u','')
treatments_clean.dose_end=treatments_clean.dose_end.str.replace('u','')

treatments_clean.dose_start=treatments_clean.dose_start.astype('int64')
treatments_clean.dose_end=treatments_clean.dose_end.astype('int64')

#### Test

In [55]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phonenumber     491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [56]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Define

- `patients` table: multiple number formats

Settle on 1234567890 format by stripping all non-digit characters, and pad all 10 digit numbers with a leading 1 (for country code)

#### Code

In [57]:
patients_clean.phonenumber.head()


0        951-719-9170
1    1 (217) 569-3204
2        402-363-6804
3    1 (732) 636-8246
4        334-515-7487
Name: phonenumber, dtype: object

In [58]:
patients_clean.phonenumber=patients_clean.phonenumber.str.replace(r'\D+','').str.pad(11, fillchar='1')

#### Test

In [60]:
patients_clean.phonenumber.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phonenumber, dtype: object

#### Define

- `patients` table includes defaul "John Doe" records

remove all records with surname "Doe"

#### Code

In [66]:
patients_clean=patients_clean[patients_clean.surname!='Doe']

#### Test

In [67]:
patients_clean.surname.value_counts()

Jakobsen             3
Taylor               3
Cabrera              2
Johnson              2
Berg                 2
Hueber               2
Woźniak              2
Ogochukwu            2
Parker               2
Lâm                  2
Lund                 2
Batukayev            2
Dratchev             2
Correia              2
Kadyrov              2
Tạ                   2
Silva                2
Cindrić              2
Grímsdóttir          2
Collins              2
Schiavone            2
Liễu                 2
Aranda               2
Lương                2
Nilsen               2
Gersten              2
Tucker               2
Bùi                  2
Souza                2
Kowalczyk            2
                    ..
Mueller              1
Harold               1
Mobourne             1
Yao                  1
Hermansen            1
Scheltens            1
Yudina               1
Cain                 1
Brodahl              1
Kristiansen          1
Cumpston             1
de Keizer            1
Skarphéðins

In [68]:
patients_clean.address.value_counts()

648 Old Dear Lane             2
2778 North Avenue             2
2476 Fulton Street            2
2403 Ripple Street            1
4271 Cherry Ridge Drive       1
2335 Webster Street           1
1990 Spring Avenue            1
1463 Beechwood Avenue         1
3402 Kildeer Drive            1
1846 Joseph Street            1
2970 Forest Avenue            1
570 Alpha Avenue              1
98 Freshour Circle            1
4155 Raccoon Run              1
4649 Joanne Lane              1
541 Green Hill Road           1
4273 Deans Lane               1
4853 University Drive         1
1540 Overlook Drive           1
4220 Simpson Square           1
2931 Romano Street            1
1733 Blackwell Street         1
1403 Clousson Road            1
2691 Kessla Way               1
4458 Stark Hollow Road        1
2458 Broadway Avenue          1
1051 Bassell Avenue           1
2386 Linda Street             1
1975 Holden Street            1
1262 Deans Lane               1
                             ..
664 Lyon

#### Define

- duplication of jakob/jake jakobsen, patrick/pat gersten, and sandra/sandy taylor. Need to remove correct one.

Determine which record is the duplicate (without a corresponding treatment patient_id), and remove from the patients table.

#### Code

In [80]:
patients_clean[patients_clean.surname.isin(['Jakobsen','Gersten','Taylor'])&patients_clean.address.duplicated(keep=False)]

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
97           98         male    Patrick   Gersten   2778 North Avenue   
131         132       female     Sandra    Taylor  2476 Fulton Street   
282         283       female      Sandy    Taylor  2476 Fulton Street   
502         503         male        Pat   Gersten   2778 North Avenue   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
29   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
97          Burr    NE    68324  United States 1954-05-03   138.2      71   
131     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
282     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
502         Burr    NE    68324  United States 1954-05-03   138.2      71   

      bmi  phonenumber                      email  
24   24.4  18458587707  JakobCJakobsen@einrot.com  
29   24.4  18458587707  JakobCJakobsen@einrot.com  
97   19.3  14028484923   PatrickGersten@rhyta.com  
131  35.4  13044382648   SandraCTaylor@dayrep.com  
282  35.4  13044382648   SandraCTaylor@dayrep.com  
502  19.3  14028484923   PatrickGersten@rhyta.com

In [78]:
treatments_clean[treatments_clean.patient_id.isin([25,30,98,132,283,503])]

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
115         7.84       7.49          0.35   auralin          51        58   
347         7.96       7.51          0.45   novodra          28        26   

    adverse_reaction  patient_id  
115              NaN         132  
347     hypoglycemia          25

interestingly patrick has no treatment record. I'll remove the later duplicate.

In [85]:
patients_clean=patients_clean[~patients_clean.patient_id.isin([30,283,503])]

#### Test

In [87]:
patients_clean[patients_clean.surname.isin(['Jakobsen','Gersten','Taylor'])]

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
97           98         male    Patrick   Gersten   2778 North Avenue   
131         132       female     Sandra    Taylor  2476 Fulton Street   
426         427         male    Rogelio    Taylor  4064 Marigold Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
97          Burr    NE    68324  United States 1954-05-03   138.2      71   
131     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
426        Miami    FL    33179  United States 1992-09-02   186.6      69   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi  phonenumber                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
97   19.3  14028484923      PatrickGersten@rhyta.com  
131  35.4  13044382648      SandraCTaylor@dayrep.com  
426  27.6  13054346299    RogelioJTaylor@teleworm.us  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

#### Define

- Zaitseva weight is in kg not lbs

convert weight to lbs by multiplying by 2.20462

#### Code

In [91]:
patients_clean.loc[patients_clean.surname == 'Zaitseva', 'weight']=patients_clean.loc[patients_clean.surname == 'Zaitseva', 'weight']*2.20462


#### Test

In [92]:
patients_clean.weight[patients_clean.surname == 'Zaitseva']

210    107.585456
Name: weight, dtype: float64

In [93]:
patients_clean.weight.describe()

count    494.000000
mean     173.515153
std       33.805144
min      102.100000
25%      148.825000
50%      174.450000
75%      199.725000
max      255.900000
Name: weight, dtype: float64

### Template workflow

#### Define


#### Code

#### Test